### DISCUSSION TITLE
ATM90E26 and ATM90E3x Energy Monitor integration to Tasmota

### DISCUSSION THREAD

**Discussion by DitroniX** [arendst/Tasmota#17801] [Q&A]
Hopefully this is the right place to ask...

Designed an Energy Monitor board/SDK, based on an ESP32 and ATM90E26.  Whilst this works well with Arduino or VSCode/PIO, would be great to integrate with Tasmota, in order to expand integration for Tasmota users.

The question is, having trawled, cannot see anything jump out, so was wondering if anyone knows of a Tasmota integration to this series of Energy Monitor devices?  If not, will investigate custom Tasmota ATM90Ex code.

Many thanks
Dave

**Comment by barbudor** [comment]
I don't see this part in the list of supported devices
Have you implemented the UART or SPI interface ?
There are a few existing drivers for UART based devices such as the BL09XX family
There is 1 use-case of a SPI energy measurment device : ADE7953

Out of curiosity, may I ask what is your objective with this board ?
Will you resell it ? Or is it intended to be OSHW ?

**Comment by DitroniX** [comment]
Hi, have implemented SPI and other GPIO is also used (project details hackster.io/ditronix/gtem-esp32-atm90e26-energy-monitor-board-sdk-33de21)  UART option is included on the board too, but not the default.

My requirements were to effectively provide a reliable means for import/export monitoring, the old CT clamps and ESP8266, just did not stack up.  These monitoring the tails and a number of solar/battery SUN GTIL2000 GTIL2 inverters. The ATM90E26, with ESP32, works brilliantly, is so accurate and reliable, providing the stats I need.  

The board was developed for my use and interest, and then grew, see github.com/DitroniX/GTEM-Grid-Tie-Energy-Monitor),  Have written a bunch of test code originally in Arduino IDE and then expanded in VSCode.  This is working well, talking to my Domoticz and other MQTT brokers.  However, would be great to include in Tasmota - hence my query.

A three phase version is being developed, so I can monitor multiple inverters etc. with one ESP32.  Plan to use the ATM90E32 or ATM90E36.  Will make this available too in a few weeks.

Regarding the GTEM board itself, have published details of it and make the PCB available.  However, for ease, and due to the compact nature of the board in 0603, those I have spoken too prefer a tested board.

Appreciate your comments.  Will probably end up playing with the Tasmota firmware in VSCode and add the library of functions and calibration code from my existing code.   Ideally need inclusion of the EEPROM routines too, as this makes it so much easier when calibrating.

Happy to send a board, or two, out to anyone that feels they could help with this project.

Hope this helps
Dave

**Reply by barbudor** [reply]
Thanks for the details. Indeed nice products.
If your 3 phase boards is price-competitive with Shelly 3EM, I could interrested by purchasing one.

**Comment by DitroniX** [comment]
Interesting and appreciate your comments.. Just took a quick peep at the _Shelly 3EM_ basics, and it looks to be an aesthetically neat and compact unit, that is designed to fit on the DIN rail inside the consumer unit.  

Clearly a lot of work, development and tooling costs went into the commercial design, compactness, safety, Phone app etc. Although it shows a CE DoC, strangely could not see an UKCA DoC or FCC certification listing. https://fccid.io/2ALAY, so I do not know how new this is to the market?, so maybe not commercially approved yet for the UK, USA etc.  That aside, it looks good and worth investigating.

My thoughts on the [IPEM 3 phase SDK kit](https://github.com/DitroniX/IPEM-IoT-Power-Energy-Monitor) (or rather, 3 x single phase monitoring), is in concept at the moment,  Have some ideas but would like it to ideally fit into an optional small enclosure, same as existing GTEM does.  The reason being, most consumer units are rather 'busy' and no real space inside, so an external design to safely house the board, certainly for me, is the route.  Also, CT clamps are rather chunky, so yet again, this is to be considered and will also test the smaller bore options.

What I am doing is playing with concepts and whilst the technology is proven, design and manufacture (whether for the community or populated PCB, needs to be right (my OCD).   Initially, have taken my GTEM design and 'virtually' expanded for three CT inputs, as seen below, to visualise it - also keeping the ESP32.  However, the current board size, even with 0603, will be quite 'busy' even without all parts included, and the ATM90E32 populated. That said, some parts will shrink, such as the Xtal.   Ideally do not want to go double sided SMT.  So I am pondering at the moment, even if it means going to 0402, or 0201, size passives.  This then also makes it much more difficult to manually populate.

![GTEM-3 ESP32 ATM90E32 SDK 2301-001 Concept Only](https://user-images.githubusercontent.com/20687856/215255640-aa8b37ac-95f3-48db-8a4a-2f16d08387d9.png)

As for selling cost, trying to keep as low a sensibly possible, it is expected to be around £50.  Once I have pinned down the final BOM, pricing will become clearer.  As with all my community SDK boards, the development is self funded but that said, much professional R&D goes into them, whether for personal use, or made available to STEM makers.

_Now is a good time to let me know if any options or ideas can be included....._   

Anyway, enough waffling.
Dave

PS. Just found this link on Shelly, which will peruse further https://github.com/arendst/Tasmota/discussions/13515

**Reply by barbudor** [reply]
Living in EU, the CE standard is enough for me ;) Even when purchasing devices from Ali 🙄 

Shelly chosed RJ connectors while you are using jacks. Which I believe are the standard connector provided with off-the-self CT.
There choice is fine as they provide both device + CT, and yours makes sense as the CT needs to be purchased separately, or at least got off-the-self without customization.

I don't have much ideas to share, you seems to have already thought about everything 😄 
May be regarding the choice of the ESP32 module : Having the ability to install an external antenna could be valuable when the location for those boards is in the basement or not well located vs the AP.

I see you have GPIO/Sensors extensions but only with 1 single signal. Why not having 2 GPIOs to support I2C sensors ? With a connector such as Adafruit STEM or Seed Grove.

Let me know if I can be of any help.
I've not much been involved in Energy drivers (except for the BL09XX). Most of them have been desgned by Theo himself.

**Comment by DitroniX** [comment]
As for the EU, the UK should never have gone the route taken but that is now consigned to political history - to which we are paying for..

The RJ route is a interesting thought but not as straight forward as 3.5mm jacks.  Feel will keep with jacks but add solder pads under the board, so giving more flexibility.   The smaller bore CTs generally are open ended wires though, will will have think!

Valid point on the ESP32 with integral antenna.  As an RF engineer and having seen user hacks to extend antenna from UFL connectors, taking this route can be equally problematic.  This also can generate complication with compliance.   That said, it would not complicate population and testing of the boards, so will consider this - even if it is internal or external antenna variants. Thanks.

The GTEM board does have an OPTO, USER and I2C interface.  The USER providing two GPIOs.  As I need to 'slim' the interfaces for space, exposing I2C with extra GPIOs on the edge would be good. Plan to keep the central I2C, if possible, as the ability to just plug in an OLED is great. Equally, the ATM_I2 interface is not used by default, so this could be exposed bottom side.  Room for thought.

Looking at the Adafruit STEM, or Seed Grove, I see they seem to use a JST PH 4 pin header and a good heads up.  I have used the smaller JST SH 6 pin on other boards, and as they are compact and importantly SMT, it could be an option too. Will see how the top side real-estate pans.

**Comment by DitroniX** [comment]
The layout of the new 3 phase ATM90E32 and ATM90E36 IPEM IoT Power Energy Monitor board is complete and beta PCBs ordered, testing planned during February, launch March.  

Further details are being added to [https://github.com/DitroniX/IPEM-IoT-Power-Energy-Monitor  ](https://github.com/DitroniX/IPEM-IoT-Power-Energy-Monitor )

**Reply by barbudor** [reply]
Hi @DitroniX 
I see the boards are ready for back order now, or even production ?
When not sure if need the E32 or E36 is it always safer to take the E36 ?
Thanks

**Reply by DitroniX** [reply]
Hi @barbudor 

Indeed, boards have already been ordered and awaiting the first shipment, anytime.  

Regarding E32 v E36, For most parts, E32 will be perfectly fine for the majority of applications.  For this reason and to save some costs, the first shipment will be for E32.

What I have done post beta board tweaks, is add the ability for CT4 to be directed (via solder jumpers), either to the E36 I4, OR to the ESP32 ADC (in the case of either E32 or E36).  This makes better sense and adds more flexibility.

So, with the E32, you will have 3 x phase, or 3 x single phases, being monitored via the ATM90Ex high resolution 16 bit ADCs and one separate phase, to the ESP32 using the lower resolution 12 bit ADC.  In all instances, three separate mains voltage inputs are available - if needed.

Hope this helps.

**Reply by DitroniX** [reply]
You can see the separate CT inputs in this 3D example, with theE32 fitted.

![IIPEM 3D](https://camo.githubusercontent.com/36881176567fafe1405c4c3af3af6ec2f7f47f8257759b0db1eb801bf0455d79/68747470733a2f2f646974726f6e69782e6e65742f77702d636f6e74656e742f75706c6f6164732f323032332f30332f4950454d2d45535033322d41544d39304533322d41544d39304533362d53444b2d312e323330322e323031502d33442d4d6f64656c2d76322d31303234783730382e706e673f7261773d74727565)

**Reply by DitroniX** [reply]
The finished 3-Phase, 3 x single phase, or split phase boards in stock - the board offers all options as standard - with 1, 2, or 3 voltage AC inputs..  Have [published ](https://github.com/DitroniX/IPEM-IoT-Power-Energy-Monitor)my bring-up code which combines both **ATM90E32** and **ATM90E36** to provide a number of data outputs.  See Domoticz code for examples.  These variables/values would be great to integrate within Tasmota as supported devices.

![IPEM ESP32 ATM90E32 ATM90E36 Variant SDK 1 2303 202P Kit](https://user-images.githubusercontent.com/20687856/229873996-9a2f1c3c-17b3-49a3-9d0c-7f571561ea2a.jpg)

![IPEM ESP32 ATM90E32 ATM90E36 SDK 1 2303 202P Sealed](https://user-images.githubusercontent.com/20687856/229874497-f43ba780-e6b8-4f3d-8304-1d517418599f.jpg)
